### ISSUE TITLE
Notice: Undefined property: DateTime::$date

### ISSUE THREAD

**Issue by peternijssen** [myclabs/DeepCopy#38]
Since I updated to PHP 7.0.7 I have an issue duplicating objects with a DateTime property. Every version before 7.0.7 worked without any issues.

the error I am getting is;
`Notice: Undefined property: DateTime::$date`

The error is thrown at this line:
`$propertyValue = $property->getValue($object);`
https://github.com/myclabs/DeepCopy/blob/a8773992b362b58498eed24bf85005f363c34771/src/DeepCopy/DeepCopy.php#L179

there are older bug reports from PHP regarding this issue, so I am unsure if this library should be improved or that this is just PHP related;
http://stackoverflow.com/questions/14084222/why-cant-i-access-datetime-date-in-phps-datetime-class-is-it-a-bug

Also see my try on 3v4l to isolate the issue;
https://3v4l.org/NOGDr

**Comment by mnapoli** [comment]
That's interesting! The best solution I think would be to not try to iterate on internal classes properties: objects like `DateTime` or `stdClass` should simply be cloned as-is I guess (because they usually don't have references to other objects so we can treat them as values).

**Comment by Vesax** [comment]
Also `Notice: Undefined property: DateTimeZone::$timezone_type`

**Comment by tchiotludo** [comment]
The same bug here in php 7.0.7 cloning a Carbon object with this error : 
```
Unknown getter 'date'
InvalidArgumentException code(0) in /vendor/nesbot/carbon/src/Carbon/Carbon.php line 527

at Carbon\Carbon->__get ('date') in Carbon.php line 527
at getValue (Object(Cawa\Date\DateTime)) in DeepCopy.php line 179
```

In case it may help you, I've a similar bug on datetime related to this thread : 
http://news.php.net/php.internals/93826%20view%20original
Datetime is not accessible via reflection and it seems to be "normal"

In my case, for a recursive serialize of object, I used a ugly hack that handle datetime like undeclared public properties : https://github.com/cawaphp/cawa/blob/master/src/Serializer/Serializer.php#L128
and unserialize is also hacky for internal class : https://github.com/cawaphp/cawa/blob/master/src/Serializer/Serializer.php#L205

@mnapoli : for simple cloning, take care about extends class that will contains reference like DatePeriod for example

**Comment by peternijssen** [comment]
@mnapoli I added a check for internal objects. (see pull request #42) The problem is now that multiple tests are failing because the tests are depending on the StdClass object. I am not sure if we should exclude that class from cloning or not.

The only thing I can think off right now is to let StdClass be cloned or change the tests to actually have a self defined object.

**Comment by ladamalina** [comment]
@theofidry Please, merge your fix asap. Our Jenkins unit tests are stuck because of DeepCopy bug with DateTime::$date 🙇  🙇

**Comment by theofidry** [comment]
@ladamalina I'm not a maintainer and I don't have write access either, so the maintainers need to review my fix first and decide to merge it or not :)

**Comment by mnapoli** [comment]
I'm very sorry for the delay everyone. I've merged #43 and tagged [1.5.2](https://github.com/myclabs/DeepCopy/releases/tag/1.5.2), please check out if all works fine for you now.

Thanks @theofidry and others who contributed and review the fix.

**Comment by theofidry** [comment]
You're welcome, works like a charm on my end ;)
